distribution/tools/entitlement-agent/src/main/java/org/elasticsearch/entitlement/agent/EntitlementAgent.java

```diff
@@ -9,13 +9,66 @@
 
 package org.elasticsearch.entitlement.agent;
 
-import org.elasticsearch.entitlement.runtime.api.EntitlementChecks;
+import org.elasticsearch.core.SuppressForbidden;
+import org.elasticsearch.entitlement.api.EntitlementChecks;
+import org.elasticsearch.entitlement.instrumentation.InstrumentationService;
+import org.elasticsearch.entitlement.instrumentation.MethodKey;
 
+import java.io.File;
+import java.io.IOException;
 import java.lang.instrument.Instrumentation;
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.Set;
+import java.util.jar.JarFile;
 
 public class EntitlementAgent {
 
     public static void premain(String agentArgs, Instrumentation inst) throws Exception {
-        EntitlementChecks.setAgentBooted();
+        // Add the bridge library (the one with the entitlement checking interface) to the bootstrap classpath.
+        // We can't actually reference the classes here for real before this point because they won't resolve.
+        var jarsString = System.getProperty("es.entitlements.bridgeJars");
+        if (jarsString == null) {
+            throw new IllegalArgumentException("System property es.entitlements.bridgeJars is required");
+        }
+        addJarsToBootstrapClassLoader(inst, jarsString);
+
+        Method targetMethod = System.class.getMethod("exit", int.class);
+        Method instrumentationMethod = EntitlementChecks.class.getMethod("checkSystemExit", Class.class, int.class);
+        Map<MethodKey, Method> methodMap = Map.of(INSTRUMENTER_FACTORY.methodKeyForTarget(targetMethod), instrumentationMethod);
+
+        inst.addTransformer(new Transformer(INSTRUMENTER_FACTORY.newInstrumenter("", methodMap), Set.of(internalName(System.class))), true);
+        inst.retransformClasses(System.class);
+    }
+
+    @SuppressForbidden(reason = "The appendToBootstrapClassLoaderSearch method takes a JarFile")
+    private static void addJarsToBootstrapClassLoader(Instrumentation inst, String jarsString) throws IOException {
+        String[] jars = jarsString.split(File.pathSeparator);
+        if (jars.length != 1) {
+            throw new IllegalArgumentException(jarsString + " must point to a single jar file");
+        }
+        for (var jar : jars) {
+            // System.out.println("Adding jar " + jar);
+            inst.appendToBootstrapClassLoaderSearch(new JarFile(jar));
+        }
+    }
+
+    private static String internalName(Class<?> c) {
+        return c.getName().replace('.', '/');
     }
+
+    private static final InstrumentationService INSTRUMENTER_FACTORY = locateInstrumenterFactory();
```

Comment:
Since we are using the embedded impl plugin, this should use ProviderLocator...I'm not sure how it would work otherwise? Perhaps something trappy about the test classpath?

Reply:
Oh! Yeesh, that sounds bad.